### DISCUSSION TITLE
Component that uses pinia in library mode can't be imported and reused in another app

### DISCUSSION THREAD

**Discussion by leopsidom** [vuejs/pinia#1073] [Help and Questions] [answered]
### Reproduction

I created a simple counter component that uses pinia to manage state: https://github.com/leopsidom/pinia-lib and build it in `lib` mode with vite and published  it as an npm package here: https://www.npmjs.com/package/pinia-lib

Then I try to import the package from another Stackblitz app here:  https://stackblitz.com/edit/github-yjj6ch?file=src/App.vue

### Steps to reproduce the behavior

1. Go to [stackblitz app](https://stackblitz.com/edit/github-yjj6ch?file=src/App.vue)
2. wait for page to load and open the browser console.

### Expected behavior

The component should work

### Actual behavior

It throws a pinia error when importing the component from anther app

> pinia-lib.es.js:276 Uncaught TypeError: Cannot read properties of undefined (reading '_s')

**Comment by posva** [comment, marked as answer]
you need to mark `pinia` as an external module in the bundling config. For vite that should be `build.rollupOptions.external: ['pinia']`

- https://vitejs.dev/config/#build-rollupoptions

**Reply by leopsidom** [reply]
Awesome. That works, thanks!

**Reply by depressedX** [reply]
So can I say it's recommended to use `pinia` as an external dependency in library mode? Because I just bundled `pinia` up with my **Vue2** library and it worked fine in another **vue2** app. So I'm not sure whether there be any potential problems.

**Reply by johnwc** [reply]
Over 2 years later, and this still worked.

**Comment by yaolunmao** [comment]
Is there any other way to avoid introducing `pinia`into new projects

**Comment by koyadume** [comment]
I encountered same issue today and can confirm that this is not working even after marking `pinia` as external module as suggested in https://github.com/vuejs/pinia/discussions/1073#discussioncomment-2196098 comment.

**Reply by tony-nz** [reply]
Hey @koyadume, did you manage to figure this out?

**Reply by koyadume** [reply]
Hi @tony-nz , Yes I figured it out finally. 

I had to pass pinia explicitly to various useXYZStore() pinia stores inside <script setup> of a component. For some reason it was unable to locate pinia inside a component even though I have registered it with Vue app. Here are the changes  I did to make it work - 

main.js
```
const pinia = createPinia()
app.use(pinia)
window.pinia = pinia
```

component
```
<script setup>
const xyzStore = useXYZStore(window.pinia)
</script>
```

**Reply by andresstbn** [reply]
Hi @koyadume 
Have you been able to solve this without using "window.pinia"?

**Reply by renyuneyun** [reply]
@andresstbn I encountered this today as well, and the solution @koyadume posted did not work for me. Not sure if my case is the same as yours, but just in case useful, here it is.
I did some further attempts by creating a simple counting store (in my library), and use it in my App. That shows Pinia was correctly picked up. It turned out that I need to add all other external libraries to my `build.rollupOptions.external: `, because they are both used in my library and my App. Interesting, but it worked.
(For me, I do not need to use the trick @andresstbn did.)

**Comment by webmint** [comment]
@posva Hello. I've catched error `TypeError: can't access property "_s", s is undefined` on app in production mode.  This app is using UI library, which is actually using Pinia (and its externalised). 

So I've tried to create simple store from examples from pinia docs. Result was same - no error in dev mode but error in production mode. Version of pinia `2.1.0`. But it was working. So i did changed version to `2.0.18` - and its worked. No error. 

 I think there is something wrong in `2.1.0` and higher. I cant confirm now that this error could appear from 0.19 to 0.36 - will check that tomorrow.

**Reply by renyuneyun** [reply]
Hmm, in my case, `2.0.18` did not work either, but it did change the error message. With `2.0.18`, the error is:
```
Error: [🍍]: getActivePinia was called with no active Pinia. Did you forget to install pinia?
	const pinia = createPinia()
	app.use(pinia)
This will fail in production.

    o pinia.esm-browser.js:1668
```

In all other versions I tested, it is `TypeError: can't access property "_s", s is undefined` (from `2.0.16` to `2.1.4`, selectively; both lib and app use the same version, otherwise sometimes compilation/type error).
And in all versions, running in dev mode works without issues.

In addition, I can only test down to `2.0.16`. `2.0.15` will result in compilation error.

**Reply by renyuneyun** [reply]
<del>To me, the error of `_s` appeared after upgrading to TypeScript 5(.1.6) (and `@vue/tsconfig` to `0.4.0`).
When I was with TypeScript 4 (with `@vue/tsconfig` `0.1.3`), the above marked answer was the key thing I needed. After changing config, Pinia worked without issues.</del>

No. In fact, this happens only when I'm using local version of the library (with `npm link`). If I publish the library to npm, and then use it, the error of `_s` does not happen.
That's weird, because just a few days ago I managed to correctly use the local library (I'm new to splitting a JS/TS library, and that was my first successful attempt). It was working correctly IIRC.

**Reply by Flyingdot** [reply]
@renyuneyun I have exactly the same problem. Have you found a solution?

**Reply by MahmoudH96** [reply]
I'm also facing the same problem in production, when I want to use pinia from a shared library, I'm getting the same error "Cannot read properties of undefined (reading '_s')"

**Reply by alfonsovgs** [reply]
I have the same error.

**Reply by renyuneyun** [reply]
@Flyingdot Hi. In my case, it turned out to be related to Quasar's static compilation, which in turn was because of my incorrect way of getting the store. See https://github.com/freddy38510/quasar-app-extension-ssg/issues/379.
Although, as you can see, the issue was not closed, because I'm still encountering some issues with the suggested solution.

**Reply by botmaster** [reply]
Same issue here.
Any workaround ?

**Reply by liorchamla** [reply]
Bumping here since externalising with Rollup did not help, here is my vite configuration : 
```
lib: {
  entry: resolve(__dirname, "src/index.ts"),
  name: "SharedComponents",
  formats: ["es", "umd"],
  fileName: (format: string) => `shared-components.${format}.js`,
},
rollupOptions: {
  // Exclude dependencies from the bundle
  external: ["vue", "pinia"],
  output: {
    globals: {
      vue: "Vue",
      pinia: "Pinia",
    },
  },
},
```

This works very well on dev mode but get the "_s" issue once builded ...

**Reply by itcrdd** [reply]
spend the whole day to get a fix for this  trying rollupOptions with vite but with no success , can confirm solution from [koyadume](https://github.com/koyadume), my solution was:

`app.use(pinia);
setActivePinia(pinia);
(window as any).pinia = pinia`

and then on state of my pinia store add the pinia store from shared package: 
`baseStore: useBaseStore((window as any).pinia)`

**Reply by liorchamla** [reply]
Thank u, but I don't think this could work on SSR since `window` won't exist :'( anyway, I deleted Pinia from my librarie, a composable was good enough

**Reply by Eva-F** [reply]
I also ran into this issue (Vue 3 + Pinia) while creating a custom component and was looking for a solution.
In the end, debugging helped me.
The error was caused by my store file, in which I was referencing another store file—something like this:
```
import {useAuthStore} from "@/stores/auth.store.ts";
const authStore = useAuthStore();
export const useNotificationStore = defineStore('notification', {
...
}
```
The solution involved using authStore as a local.
